### PR TITLE
feat(broccoli): add incremental dartfmt plugin

### LINKED ISSUE #2208: Move dart formatting to incremental plugin in build pipeline

### PR DESCRIPTION
**Pull request by caitp**
Closes #2208

### CODE REVIEW

**Comment by caitp** [comment]
PTAL

**Comment by caitp** [comment]
well, one of the integration tests fail, no idea whats breaking it. Maybe one of the dart-y people know

File: tools/broccoli/broccoli-dartfmt.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
+/// <reference path="../typings/node/node.d.ts" />
+/// <reference path="../typings/fs-extra/fs-extra.d.ts" />
+import fse = require('fs-extra');
+import path = require('path');
+import {wrapDiffingPlugin, DiffingBroccoliPlugin, DiffResult} from './diffing-broccoli-plugin';
+var Q = require('q');
```
Comment (IgorMinar) [inline tools/broccoli/broccoli-dartfmt.ts]: can we use native promises instead? they are already available in node 0.12

File: tools/broccoli/broccoli-dartfmt.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+class DartFormatter implements DiffingBroccoliPlugin {
+  static includeExtensions = ['.dart'];
+
+  private DARTFMT: string;
+  private verbose: boolean;
+
+  constructor(public inputPath: string, public cachePath: string, options) {
+    if (!options.dartSDK) throw new TypeError("Missing Dart SDK");
```
Comment (IgorMinar) [inline tools/broccoli/broccoli-dartfmt.ts]: Why `TypeError`? Shouldn't it be just an `Error`?

Comment (caitp) [inline tools/broccoli/broccoli-dartfmt.ts]: TypeError is the general runtime error, but it doesn't really matter. Error it is

File: tools/broccoli/broccoli-dartfmt.ts (hunk 1 above)
Comment (caitp) [inline tools/broccoli/broccoli-dartfmt.ts]: Sure

File: gulpfile.js [hunk 3]
```diff
@@ -427,7 +427,6 @@ gulp.task('!test.unit.js/karma-run', function(done) {
 gulp.task('test.unit.dart', function (done) {
   runSequence(
     'build/tree.dart',
-    'build/format.dart',
```
Comment (IgorMinar) [inline gulpfile.js]: shouldn't we also remove the `build/format.dart` gulp task?

Comment (caitp) [inline gulpfile.js]: done

**Comment by caitp** [comment]
something seems to have broken the watch plugin here :(

**Comment by IgorMinar** [comment]
watch plugin? can you be more specific?

**Comment by IgorMinar** [comment]
otherwise looks good, except for the broken tests. I asked @vsavkin to take a look.

**Comment by caitp** [comment]
> watch plugin? can you be more specific?

It looks like the same issue as https://github.com/ember-cli/ember-cli/issues/2226 so I'm hoping a reboot fixed it

**Comment by IgorMinar** [comment]
can you check your tmp directory and make sure that it's empty or nonexistent while the build is not running?

**Comment by caitp** [comment]
The tmp directory gets cleaned up properly, it was just a random issue with FSEvents (re: the watcher being broken). A reboot seems to have fixed it, so I blame Apple stuff.

**Comment by vsavkin** [comment]
I looked into it. The build fails because one of the tests cannot find the `simple_annotation_files/expected/bar.ng_meta.json` file. It looks like this is the only JSON (not *.dart) file required used in the test. So maybe the transformation filters out non-dart files?

**Comment by caitp** [comment]
ah I see. I think this needs to change `getSourceTrees()` itself to only get at the dart files

**Comment by IgorMinar** [comment]
@caitp instead of changing the sourceTree, how about ignoring the non-dart files and passing them through by default?

I think this might be an architectural question for all plugins, but for now, I would just relink the inputs and outputs for all non .dart files

**Comment by caitp** [comment]
that's the approach taken here currently. I think it will probably work, but it's not "ideal" I guess

**Comment by caitp** [comment]
looks like it worked. LGTY?

**Comment by IgorMinar** [comment]
yeah. lgtm. thanks!

**Pull request closed without merging**